### DISCUSSION TITLE
Different background colours for WindowName widget for when there are windows or not

### DISCUSSION THREAD

**Discussion by hamensman** [qtile/qtile#2867] [Ideas] [answered]
Maybe a bit random, but thought about for the `WindowName` widget if you can set a background colour for when there are _no_ windows, and another for when there _are_ windows. Essentially a highlight. And potentially, users could change how this looks based on their preferences, like in the `GroupBox` widget.

Something like this:

![Screenshot_2021-10-09_14-51-42](https://user-images.githubusercontent.com/76445071/136660894-3741b5db-a0cc-4c6d-89d5-c4fd5e15f5d2.png)
![Screenshot_2021-10-09_14-52-10](https://user-images.githubusercontent.com/76445071/136660897-86054dbb-52dd-442c-8c33-3e79b7fae6d0.png)

Please tell me your thoughts.

**Comment by elParaguayo** [comment, marked as answer]
Again, no reason you can't do it. Probably an easier change than your previous suggestion!

You just need some logic to set the colour used by `self.drawer.clear` in the `draw` method.

**Reply by hamensman** [reply]
Are the source code files stored locally somewhere in the user's computer? I'm on Arch Linux.

**Reply by CSpencerND** [reply]
`/usr/lib/python3.9/site-packages/libqtile/widget/windowname.py`
I'd recommend copying the file to `~/.config/qtile/custom_windowname.py` so you don't mess with the default stuff. It's safer. Then in your config, import the custom version - `import custom_windowname` and use it instead of the original one.

**Reply by elParaguayo** [reply]
Nooo. Don't edit those!

Clone the GitHub repo and edit that.

**Reply by CSpencerND** [reply]
Nah, I wasn't saying to edit the actual source files. Just to make a copy, but yeah, cloning does make more sense. This illustrates why finishing a course on a programming language is just the beginning of the learning process.

**Reply by elParaguayo** [reply]
Cloning is the only way to make changes and submit pull requests so good to learn this.

**Reply by hamensman** [reply]
> You just need some logic to set the colour used by `self.drawer.clear` in the `draw` method.

Like in the `GroupBox` widget then?

**Reply by elParaguayo** [reply]
`WindowName` subclasses `libqtile.widget.base._Textbox` so it doesn't actually have it's own `draw` method. Instead, it uses the method from `_Textbox`:
https://github.com/qtile/qtile/blob/7d2e9b1cc14b1f7ed7bf9e8627e004f725d998f0/libqtile/widget/base.py#L432-L440

The bit that sets the background is this line:
https://github.com/qtile/qtile/blob/7d2e9b1cc14b1f7ed7bf9e8627e004f725d998f0/libqtile/widget/base.py#L435

So you would need some logic to choose the background colour depending on whether the widget has text or not.

**Reply by hamensman** [reply]
Let's say I have `custom_base.py`. When I want to inherit that in `custom_windowname.py`, do I leave `from libqtile.widget import base` as it is, or do I need to rewrite that to something like `import custom_base`?

**Reply by elParaguayo** [reply]
If you have a file called `custom_base.py`, put it in the same folder as `config.py` and then you can do `import custom_base`.

Also, if `custom_base.py` has a widget in it called:
``` python
from libqtile import widget

class CustomWindowName(widget.WindowName):
    ...
```

You can can `from custom_base import CustomWindowName`.

**Reply by hamensman** [reply]
> `WindowName` subclasses `libqtile.widget.base._Textbox `so it doesn't actually have it's own `draw` method

Would it be better if I somehow write a `draw` method of its own?

If I use the draw method from (custom_)`base.py`, there's only one background variable. So let's say I applied the background colour **only** if there's text. If there's no text - and then no background applied as a result- it would just be black.

Which I suppose this is why the `GroupWidget` has the `background` colour variable, but also a separate variable for the highlight colour.

*Actually, never mind. I could probably just write another background variable in `custom_base.py`?

**Reply by elParaguayo** [reply]
All you really need to do is subclass the `WindowName` widget, copy the `draw` method from `_TextBox` and then add the logic needed for the additional background colour (e.g. adding a new parameter in `defaults`). You don't need to rewrite `base` at all.

**Reply by hamensman** [reply]
This is what I did:
___

I have a file (in `~/.config/qtile`) called `custom_windowname.py.` the class name is `CustomWindowName`. I've added in a `highlighy_color` parameter in `defaults`:

`('highlight_color', None, 'Widget highlight color'),`

and added the draw method like this:
```
def draw(self):
    if not self.can_draw():
        return
    self.drawer.clear(self.highlight_color or self.bar.highlight_color)
    self.layout.draw(
        self.actual_padding or 0,
        int(self.bar.height / 2.0 - self.layout.height / 2.0) + 1
    )
    self.drawer.draw(offsetx=self.offsetx, width=self.width)
```
 See [custom_windowname.txt](https://github.com/qtile/qtile/files/7318226/custom_windowname.txt) for full context (rename to `custom_windowname.py` accordingly)
___

I have `import custom_windowname` in my config file.

I have:
```
widget.CustomWindowName(font="Noto Sans",
         fontsize = 12,
         foreground = colors[7],
         background = colors[0],
         highlight_color = colors[7],
```
in my `widgets_list`. See [hamens_config.txt](https://github.com/qtile/qtile/files/7318275/hamens_config.txt) for full context (rename to `config.py` accodingly)
___

But I've received a config error, so I've definitely missed someting:  [qtile.log](https://github.com/qtile/qtile/files/7318212/qtile.log)

Apologies for taking your time with this.

**Reply by elParaguayo** [reply]
OK, your widget probably just needs to be something like this:
``` python
from libqtile import widget


class CustomWindowName(widget.WindowName):
    """Displays the name of the window that currently has focus"""
    orientations = base.ORIENTATION_HORIZONTAL
    defaults = [
        ('highlight_color', None, 'Widget highlight color'),
    ]

    def __init__(self, **config):
        widget.WindowName.__init__(self, **config)
        self.add_defaults(CustomWindowName.defaults)

    def draw(self):
        if not self.can_draw():
            return
	if self.text:
            # Assuming "text" is empty then use the highlight colour
            background = self.highlight_color
        else:
            background = self.background or self.bar.background
        self.drawer.clear(background)
        self.layout.draw(
            self.actual_padding or 0,
            int(self.bar.height / 2.0 - self.layout.height / 2.0) + 1
        )
        self.drawer.draw(offsetx=self.offsetx, width=self.width)
```

Then, in your config, you do:
``` python
from custom_windowname import CustomWindowName
```
and you create the widget like this:
``` python
# you don't need the "widget." as it's not provided by libqtile.widget
CustomWindowName(font="Noto Sans",
         fontsize = 12,
         foreground = colors[7],
         background = colors[0],
         highlight_color = colors[7]
```

It's untested so, if you get any errors, look at the logs and paste them here.

**Reply by hamensman** [reply]
Here is the error file: [qtile.log](https://github.com/qtile/qtile/files/7324366/qtile.log)

The main thing is that `base` in `orientations = base.ORIENTATION_HORIZONTAL` is not defined.

Do you reckon the creation of `custom_base.py` would sort this out?

**Reply by elParaguayo** [reply]
Just delete the orientations line from CustomWindowName. You don't need it.

**Reply by hamensman** [reply]
Ok. So Qtile now responds when initiating a qtile restart (the other times, it did not). But I'm getting some weird results: [qtile.log](https://github.com/qtile/qtile/files/7324454/qtile.log)

Without even declaring `highlight_color` (I commented it out) in the config file, it made a huge section of the bar blank, removing my system tray and many of my other widgets like the date.

![Screenshot_2021-10-11_18-52-41](https://user-images.githubusercontent.com/76445071/136834125-07df2c7c-8226-49f6-87c0-138d13b0fcfd.png)

**Reply by elParaguayo** [reply]
Probably because I made a mistake somewhere! I'll take a look a bit later.

**Reply by elParaguayo** [reply]
`highlight_color` can't be `None` so it does need to be set for now. Can you set a value and paste the log again.

**Reply by elParaguayo** [reply]
Might be worth changing this line:
``` python
background = self.highlight_color
```
to
```python
background = self.highlight_color or self.background or self.bar.background
```

**Reply by hamensman** [reply]
[qtile.log](https://github.com/qtile/qtile/files/7324605/qtile.log)

There is no longer the crazy blank part on the bar. And I've uncommented `highlight_color` - so a value for that is set. But it doesn't show any highlight colour either.

**Reply by elParaguayo** [reply]
There were some tabs and spaces mixed up in that. Try this:
``` python
class CustomWindowName(widget.WindowName):
    """Displays the name of the window that currently has focus"""
    defaults = [
        ('highlight_color', None, 'Widget highlight color'),
    ]

    def __init__(self, **config):
        widget.WindowName.__init__(self, **config)
        self.add_defaults(CustomWindowName.defaults)

    def draw(self):
        if not self.can_draw():
            return
        if self.text != self.empty_group_string:
            # Assuming "text" is empty then use the highlight colour
            background = self.highlight_color or self.backgroun or self.bar.background
        else:
            background = self.background or self.bar.background
        self.drawer.clear(background)
        self.layout.draw(
            self.actual_padding or 0,
            int(self.bar.height / 2.0 - self.layout.height / 2.0) + 1
        )
        self.drawer.draw(offsetx=self.offsetx, width=self.width)
```

**Reply by hamensman** [reply]
[qtile.txt](https://github.com/qtile/qtile/files/7324734/qtile.txt) (Didn't allow me to upload as `.log` for some reason)

When I restart, part of the bar was blank again. I then switched screens. Went back to that first screen (1), and the bar looks normal again...

Except that now there's no text at all for the window name - even when windows are open.

![Screenshot_2021-10-11_20-00-41](https://user-images.githubusercontent.com/76445071/136842914-212a4f5a-3e3f-47dd-957b-6412d0ef647a.png)
![Screenshot_2021-10-11_20-01-02](https://user-images.githubusercontent.com/76445071/136842917-49530cd3-65ea-479b-be1c-8f3be8388482.png)

Here you can see my browser is open. But it doesn't show the window name on the bar:
![Screenshot_2021-10-11_20-04-10](https://user-images.githubusercontent.com/76445071/136843280-1a92c14f-ea01-41e7-bd5c-23b3bb56f8f4.png)

Ok. Just saw `self.background` was missing a 'd'. Added that in. Now the window names **do** show up. But just same issue of no highlight colour: 
[qtile.log](https://github.com/qtile/qtile/files/7324810/qtile.log)

**Reply by elParaguayo** [reply]
Sigh. My typing skills suck today... One of the `background`s is missing a d at the end.

**Reply by hamensman** [reply]
That's no problem. I'm just grateful you're helping me out with a very powerful learning experience!

Oops! Added wrong log! 
[qtile.log](https://github.com/qtile/qtile/files/7324871/qtile.log)

**Reply by elParaguayo** [reply]
Does that mean it's still now working for you? It works for me!

**Reply by hamensman** [reply]
> It works for me!

It does?! So for you, when no windows are open, it's just base bar colour? And when there are windows, there's a highlight colour?

I must have messed something up...

**Reply by elParaguayo** [reply]
Post your `conf.py` file.

**Reply by hamensman** [reply]
[hamens_config.txt](https://github.com/qtile/qtile/files/7325102/hamens_config.txt) (rename to `config.py` accordingly.

**Reply by elParaguayo** [reply]
`custom_windowname.py` too please.

**Reply by hamensman** [reply]
Here you go: [custom_windowname.txt](https://github.com/qtile/qtile/files/7325147/custom_windowname.txt)

**Reply by elParaguayo** [reply]
Don't see why that's not working.

**Reply by hamensman** [reply]
Right now the logs say just this:
`ERROR libqtile group.py:layout():L122 No such layout: modatall`

This suggests somewhere `monadtall` is spelt wrong, but not sure how relevant that is...

**Reply by elParaguayo** [reply]
That's in line 184 of your config. Last entry in the list.

**Reply by hamensman** [reply]
Now, there is nothing in the error logs. But still not working.

Can you post screenshots of what it looks like for you please?

Hmmm.... Now I have this after restarting my computer: 
[qtile.log](https://github.com/qtile/qtile/files/7325269/qtile.log)

**Reply by elParaguayo** [reply]
![20211011_214829](https://user-images.githubusercontent.com/946265/136853949-e1fb490a-7603-4494-8c60-383b2ee3339b.png)
![20211011_214922](https://user-images.githubusercontent.com/946265/136853951-a15abaf8-2238-4268-bc2d-a8b3167f4c1f.png)

**Reply by elParaguayo** [reply]
That `has_keyboard` error is unrelated.

**Reply by hamensman** [reply]
This is quite perplexing... Your's is working perfectly!

**Reply by elParaguayo** [reply]
Can you open a terminal and run the following:
```
qtile cmd-obj -o screen bar top -f info
```
I want to see if the widget is included in your list of widgets.

**Reply by hamensman** [reply]
Here you go: [info.log](https://github.com/qtile/qtile/files/7325331/info.log)

It does appear to be listed.

**Reply by elParaguayo** [reply]
Yeah. So it is there.

Try:
```
qtile cmd-obj -o screen bar top widget customwindowname -f info
```

**Reply by hamensman** [reply]
Here's what I got: [info.log](https://github.com/qtile/qtile/files/7328806/info.log)

**Reply by elParaguayo** [reply]
Ahhhh..

Look at your config. You've got a spelling mistake: `highight_color = colors[1]`

Should be: `highlight_color = colors[1]`

**Reply by hamensman** [reply]
Ahh! You're kidding me!

It works now! I feel so stupid...

Thanks a lot @elParaguayo! I gained a bit more confidence in Python with this!

**Reply by elParaguayo** [reply]
Don't feel stupid. We've all done it before (and I'll do it again, no doubt).

Glad it's working.
